OP-166 | Download MySQL during build

pviotti opened the pull request:
Download MySQL binaries during build, as it is done for JRE.  

Close #123 

**Open issues**
- [x] `Xchang32.exe` is currently used in [`oh.bat`](https://github.com/informatici/openhospital/blob/master/poh-bundle-win/oh.bat) to replace MySQL configuration settings. Such tool is not present in the last version of MySQL (and it should probably be replaced with something like a Powershell command anyway...) [will probably not fix in this PR - there's already too many changes]
 - [x] cache and parallelize downloads
 

**Future improvements** (in other PRs)
 - turn build script in Makefile so that we can skip automatically file that have already been built or downloaded
 - bump poh version (since we changed quite a lot, perhaps it's time to dare a 1.0?)


**Dev notes**
To test locally, set `export OH_VERSION=develop` since the current matching tag across repos might not build successfully. 

for MySQL 8.x these are the URLs:
 - https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-8.0.21-linux-glibc2.17-x86_64-minimal.tar.xz
 - https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-8.0.21-linux-glibc2.12-i686.tar.xz
 - https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-8.0.21-winx64.zip (there is no win32 version)

pviotti committed:
feat: download mysql 5.7.30 during build

pviotti committed:
remove mysql binaries from repo

pviotti committed:
remove mysql binaries for linux x64 bundle

pviotti committed:
remove mysql binaries for linux x32 bundle

pviotti committed:
change startup script for linux bundles

pviotti committed:
chore: git-ignore full distribution folder

pviotti committed:
chore: git-ignore file produced by maven

pviotti committed:
chore: git-ignore dump_error.log and all log files

pviotti committed:
feat: change mysql folder name in oh.bat

pviotti committed:
feat: improve mysql dump process

pviotti committed:
chore: add echo message when compiling documentation

pviotti committed:
feat: download mysql and jre in parallel

pviotti committed:
chore: mv instead of cp release files to release folder

pviotti committed:
chore: add wget and xargs to required tools list

mwithi reviewed (changes requested):
1. Please add also `mysqldump` to the requirements. It would be nice also to show the standard output from the OS, e.g.:
```
Command 'mysqldump' not found, but can be installed with:

sudo apt install mysql-client-5.7
sudo apt install mariadb-client-10.1
```

2. When building locally, it needs `rm -r core gui doc` if you want to run more the once. Shall we add this?

3. When dump_error.log contains `mysqldump: Error:` it should stop the build (warnings are allowed) :
```
mysqldump: [Warning] Using a password on the command line interface can be insecure.
mysqldump: Error: 'Access denied; you need (at least one of) the PROCESS privilege(s) for this operation' when trying to dump tablespaces
```

4. It seems the mysqldump error above can be solved with [this](https://dba.stackexchange.com/a/273040) (tested):
```
mysqldump [etc etc]  --no-tablespaces
```

5. at the moment checks are failing because it is fetching v1.10.0 (from `git describe --abbrev=0 --tags`) when tests were relying on the docker mysql instance rather than the in-memory db as it is now, so we could keep running `mvn package` before `docker-compose down`, wdyt?
Tested both: cloning from HEAD with in-memory tests and switching the two commands positions

6. if we `mv` away `*.zip` and `*.gzip` files from root folder, the `checksum` command will fail.

Thanks a lot!!!

pviotti committed:
fix: replace mysql with mysqldump in list of required tools

pviotti committed:
fix: add no-tablespaces option to mysqldump command

pviotti committed:
feat: exit on sqldump error

pviotti committed:
fix: docker-compose down after mvn package

pviotti committed:
fix: do checksum in the right folder

pviotti committed:
fix: remove core gui and doc if present before cloning

pviotti committed:
chore: improve mysql waiting loop

pviotti committed:
chore: use pushd and popd instead of cd

pviotti committed:
chore: wget avoid download if files are there already (--no-clobber)

pviotti committed:
feat: increase maven parallelism

pviotti committed:
chore: use -rf instead of || true for rm commands

pviotti committed:
fix: checksum of zip and tar.gz files

pviotti committed:
feat: add clean-downloads target to makefile

pviotti committed:
feat: optionally get version from env variable

pviotti committed:
feat: replace xchang32 with replace in oh.bat

pviotti committed:
feat: add my.ori and change oh.bat accordingly

mwithi committed:
Solved few problems on Windows startup (after build)

- [x] It gets stuck when using `replace.exe`: removed quotes and
commented out "^x5c" "^x2f" replacement
- [x] Also, `my.ori` should be searched in `cd /d %OH_PATH%`
- [x] If the command   `%OH_PATH%mysql-5.7.30-win32\bin\mysql -u root
--port=%freePort% oh < "%OH_PATH%database.sql"` fails, it should not
delete `database.sql` file
- [x] To download the mysql-package is not enough, `--initialize` should
be used first to generate the `data` folder
- [x] `my.ori` needs paths adjusting to the new mysql folder, anyway
since v5.7 there is an [Option File Processing
Order](https://dev.mysql.com/doc/refman/5.7/en/option-files.html#option-file-order)
and [Inclusion](https://dev.mysql.com/doc/refman/5.7/en/option-files.html#option-file-inclusions)
syntax, so the easiest solution I found was to move `my.cnf` to the
`mysql/bin` folder
- [x] mysqldump should not need anymore `--compatible=mysql40` option,
I've removed so we can test it
- [x] the 'oh' database is missing, so we need to create it and also the
users: adding same commands as in `step_00_create_user.sql` to the
initialization step
- [ ] probably some things apply to 'oh.sh' too

pviotti committed:
feat: use mysqldump from running container

pviotti committed:
chore: add release-files folder to gitignore

mwithi committed:
Removed useless deletion of removed file

mwithi committed:
Fixing Linux MySQL initialize and new settings for 5.7

pviotti committed:
fix: add socket file to mysql client arguments

pviotti committed:
fix: rm -rf var/lib/mysql for linux 32/64 bundles

pviotti committed:
fix: make sure /var/lib/mysql is there upon db initialization

mwithi committed:
Fixing paths in win32 bundle

mwithi committed:
fix: make sure /var/lib/mysql is there upon db initialization

improved mysql waiting loop

mwithi committed:
Excluding MySQL lib/ folder to shrink the packages

mwithi committed:
Fix: lasttag and secondlast tags retrieving and add doc changelog
